File: pipeline/reimpl/cbimport.py
"""
Reimplementation of CBIMPORT — CardDemo Batch Data Import.

Reads a multi-record export bundle produced by CBEXPORT and splits it into
separate normalized target collections (Customer, Account, Xref, Transaction,
Card).  Records with unknown type characters are logged to the error list.

Returns an ImportResult with separated lists, statistics, and an error log.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any

from .carddemo_data import (
    CustomerRecord, AccountRecord, CardXrefRecord, TranRecord, CardRecord,
)
from .cbexport import ExportRecord


@dataclass
class ImportStats:
    customers: int = 0
    accounts: int = 0
    xrefs: int = 0
    transactions: int = 0
    cards: int = 0
    errors: int = 0

    @property
    def total_imported(self) -> int:
        return self.customers + self.accounts + self.xrefs + self.transactions + self.cards


@dataclass
class ImportResult:
    customers: list[CustomerRecord] = field(default_factory=list)
    accounts: list[AccountRecord] = field(default_factory=list)
    xrefs: list[CardXrefRecord] = field(default_factory=list)
    transactions: list[TranRecord] = field(default_factory=list)
    cards: list[CardRecord] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)
    stats: ImportStats = field(default_factory=ImportStats)
    log: list[str] = field(default_factory=list)


def run_import(export_records: list[ExportRecord]) -> ImportResult:
    """Run data import — mirrors CBIMPORT PROCEDURE DIVISION."""
    result = ImportResult()
    result.log.append("CBIMPORT: Starting Customer Data Import")
    result.log.append(f"CBIMPORT: Records to process: {len(export_records)}")

    _HANDLERS = {
        "C": (_handle_customer, result.customers),
        "A": (_handle_account, result.accounts),
        "X": (_handle_xref, result.xrefs),
        "T": (_handle_tran, result.transactions),
        "D": (_handle_card, result.cards),
    }

    for rec in export_records:
        if rec.rec_type not in _HANDLERS:
            result.errors.append(
                f"CBIMPORT: Unknown record type '{rec.rec_type}' at seq {rec.seq_num}"
            )
            result.stats.errors += 1
            continue

        handler, target_list = _HANDLERS[rec.rec_type]
        parsed, err = handler(rec)
        if err:
            result.errors.append(f"CBIMPORT: Error at seq {rec.seq_num}: {err}")
            result.stats.errors += 1
        else:
            target_list.append(parsed)

    result.stats.customers = len(result.customers)
    result.stats.accounts = len(result.accounts)
    result.stats.xrefs = len(result.xrefs)
    result.stats.transactions = len(result.transactions)
    result.stats.cards = len(result.cards)

    result.log.append(f"CBIMPORT: Customers imported:     {result.stats.customers}")
    result.log.append(f"CBIMPORT: Accounts imported:      {result.stats.accounts}")
    result.log.append(f"CBIMPORT: Xrefs imported:         {result.stats.xrefs}")
    result.log.append(f"CBIMPORT: Transactions imported:  {result.stats.transactions}")
    result.log.append(f"CBIMPORT: Cards imported:         {result.stats.cards}")
    result.log.append(f"CBIMPORT: Errors:                 {result.stats.errors}")
    result.log.append("CBIMPORT: Import complete.")
    return result


# ─── Record handlers ─────────────────────────────────────────────────────────

def _handle_customer(rec: ExportRecord):
    if isinstance(rec.data, CustomerRecord):
        return rec.data, None
    return None, "data is not a CustomerRecord"


def _handle_account(rec: ExportRecord):
    if isinstance(rec.data, AccountRecord):
        return rec.data, None
    return None, "data is not an AccountRecord"


def _handle_xref(rec: ExportRecord):
    if isinstance(rec.data, CardXrefRecord):
        return rec.data, None
    return None, "data is not a CardXrefRecord"


def _handle_tran(rec: ExportRecord):
    if isinstance(rec.data, TranRecord):
        return rec.data, None
    return None, "data is not a TranRecord"


def _handle_card(rec: ExportRecord):
    if isinstance(rec.data, CardRecord):
        return rec.data, None
    return None, "data is not a CardRecord"

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbimport.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java reimplementation of CardDemo batch program CBIMPORT — Batch Data Import.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbimport.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbimport implements ProgramRunner {

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenarioName = inputs.getOrDefault("SCENARIO", "PROCESS_RECORDS").toUpperCase();
        Scenario scn = buildScenario(scenarioName);
        if (scn == null) {
            Map<String, String> err = new LinkedHashMap<>();
            err.put("error", "unknown scenario: '" + scenarioName + "'");
            return err;
        }

        // Run import
        List<String> log = new ArrayList<>();
        List<String> errors = new ArrayList<>();
        log.add("CBIMPORT: Starting Customer Data Import");
        log.add("CBIMPORT: Records to process: " + scn.records.size());

        int customers = 0, accounts = 0, xrefs = 0, transactions = 0, cards = 0;
        int errorCount = 0;

        for (ExportRecord rec : scn.records) {
            switch (rec.recType) {
                case "C": customers++; break;
                case "A": accounts++; break;
                case "X": xrefs++; break;
                case "T": transactions++; break;
                case "D": cards++; break;
                default:
                    errors.add("CBIMPORT: Unknown record type '" + rec.recType
                            + "' at seq " + rec.seqNum);
                    errorCount++;
                    break;
            }
        }

        log.add("CBIMPORT: Customers imported:     " + customers);
        log.add("CBIMPORT: Accounts imported:      " + accounts);
        log.add("CBIMPORT: Xrefs imported:         " + xrefs);
        log.add("CBIMPORT: Transactions imported:  " + transactions);
        log.add("CBIMPORT: Cards imported:         " + cards);
        log.add("CBIMPORT: Errors:                 " + errorCount);
        log.add("CBIMPORT: Import complete.");

        int totalImported = customers + accounts + xrefs + transactions + cards;

        Map<String, String> out = new LinkedHashMap<>();
        out.put("TOTAL_IMPORTED", String.valueOf(totalImported));
        out.put("CUSTOMERS", String.valueOf(customers));
        out.put("ACCOUNTS", String.valueOf(accounts));
        out.put("XREFS", String.valueOf(xrefs));
        out.put("TRANSACTIONS", String.valueOf(transactions));
        out.put("CARDS", String.valueOf(cards));
        out.put("ERRORS", String.valueOf(errorCount));
        for (int i = 0; i < log.size(); i++) {
            out.put("LOG_" + i, log.get(i));
        }
        for (int i = 0; i < errors.size(); i++) {
            out.put("ERROR_" + i, errors.get(i));
        }
        return out;
    }

    // ── Scenarios ─────────────────────────────────────────────────────────

    private Scenario buildScenario(String name) {
        switch (name) {
            case "PROCESS_RECORDS": return scenarioProcessRecords();
            case "EMPTY_INPUT":    return scenarioEmptyInput();
            case "BAD_TYPE":       return scenarioBadType();
            default: return null;
        }
    }

    private Scenario scenarioProcessRecords() {
        Scenario s = new Scenario();
        s.records.add(new ExportRecord(1, "C"));
        s.records.add(new ExportRecord(2, "A"));
        s.records.add(new ExportRecord(3, "X"));
        s.records.add(new ExportRecord(4, "T"));
        s.records.add(new ExportRecord(5, "D"));
        return s;
    }

    private Scenario scenarioEmptyInput() {
        return new Scenario();
    }

    private Scenario scenarioBadType() {
        Scenario s = new Scenario();
        s.records.add(new ExportRecord(1, "Z"));
        return s;
    }

    // ── Inner data types ─────────────────────────────────────────────────

    private static class Scenario {
        final List<ExportRecord> records = new ArrayList<>();
    }

    static class ExportRecord {
        final int seqNum;
        final String recType;

        ExportRecord(int seqNum, String recType) {
            this.seqNum = seqNum;
            this.recType = recType;
        }
    }
}
